codegen/src/PackageMetadata.scala

```diff
@@ -0,0 +1,75 @@
+package besom.codegen
+
+import besom.codegen.UpickleApi._
+
+case class PackageMetadata private (
+  name: PackageMetadata.SchemaName,
+  version: PackageVersion.PackageVersion,
+  server: Option[String] = None
+) {
+  def withUrl(url: String): PackageMetadata = {
+    val server = url match {
+      case s"https://github.com/pulumi/pulumi-${_}" => None // use default
+      case s"https://github.com/$org/$name"         => Some(s"github://api.github.com/$org/$name")
+      case _                                        => throw GeneralCodegenException(s"Unknown repo url format: ${url}")
+    }
+    PackageMetadata(name, version, server)
+  }
+
+  def toJson: String = PackageMetadata.toJson(this)
+}
+
+object PackageMetadata {
+  implicit val rw: ReadWriter[PackageMetadata] = macroRW
+
+  type SchemaName    = String
+  type SchemaVersion = String
+  type SchemaFile    = os.Path
+
+  def apply(name: SchemaName, version: SchemaVersion, server: Option[String] = None): PackageMetadata = {
+    new PackageMetadata(name, PackageVersion(version), server)
+  }
+
+  def fromJson(json: String): PackageMetadata = {
+    read[PackageMetadata](json)
+  }
+
+  def fromJsonFile(path: os.Path): PackageMetadata = {
+    read[PackageMetadata](os.read(path))
+  }
+
+  def toJson(m: PackageMetadata): String = write[PackageMetadata](m)
+}
+
+object PackageVersion {
+  type PackageVersion = PackageMetadata.SchemaVersion
+
+  private val DefaultVersion = "0.0.0"
+  private val SemverRegex =
+    """^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$""".r
+
+  def apply(version: String): PackageVersion = version.trim.stripPrefix("v") match {
+    case v @ SemverRegex(_*) => new PackageVersion(v)
+    case ""                  => new PackageVersion(DefaultVersion)
+    case v                   => throw new Exception(s"Invalid version format: $v")
+  }
+
+  def apply(): PackageVersion = new PackageVersion(DefaultVersion)
+
+  def isValid(version: String): Boolean = SemverRegex.matches(version)
+
+  implicit class PackageVersionOps(version: PackageVersion)(implicit logger: Logger) {
+    def isDefault: Boolean = version == DefaultVersion
+
+    def reconcile(otherVersion: PackageVersion): PackageVersion = {
+      (version, otherVersion) match {
+        case (version, _) if !version.isDefault => version
```

Comment:
What is the expectation if both versions are nondefault? Should the former one be taken and the latter discarded without a warning? The intent seems quite unclear to me

Reply:
would changing the name of the parameter from `otherVersion` to `defaultVersion` would make the intent more clear, or renaming the method to `versionOrDefault`?

the intent is handling missing version in schema file

Reply:
I would rather expect something like
```scala
      (version.isDefault, otherVersion.isDefault) match {
        case (true, false) => version
        case (false, true) => otherVersion
        case (true, true) => PackageVersion(DefaultVersion)
        case (false, false) => ??? // Fixme: which version to choose?
```

Reply:
```
    def reconcile(otherVersion: PackageVersion): PackageVersion = {
      (version.isDefault, otherVersion.isDefault) match {
        case (true, false) => version
        case (false, true) => otherVersion
        case (false, false) => version // we ignore the other version
        case (true, true)  =>
          // this code should be only reached in tests
          logger.warn(s"Using default version '${DefaultVersion}' for the package - suitable only for tests")
          PackageVersion()
      }
    }
    ```